<a href="https://colab.research.google.com/github/matheus97eng/Bootcamp_Alura_2021/blob/main/modulo-6/aulas/notebooks/aula_6-1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

df_clientes_cadastrados = pd.read_csv('https://github.com/alura-cursos/Avaliacao_Credito/blob/main/dados/clientes_cadastrados.csv?raw=true')
df_clientes_cadastrados.head()

ID_Cliente Genero  Tem_Carro  ...   Ocupacao  Tamanho_Familia  Anos_empregado
0     5008804      H          1  ...     Outros              2.0            12.4
1     5008805      H          1  ...     Outros              2.0            12.4
2     5008806      H          1  ...  Seguranca              2.0             3.1
3     5008808      M          0  ...      Venda              1.0             8.4
4     5008809      M          0  ...      Venda              1.0             8.4

[5 rows x 18 columns]

# Analisando as features

## 1ª hipótese: cada `ID_Cliente` representa um cliente diferente?

In [ ]:
df_clientes_cadastrados.shape

(438557, 18)

In [ ]:
df_clientes_cadastrados['ID_Cliente'].value_counts()

7603224    2
7742853    2
7836711    2
7091721    2
7022327    2
          ..
5372264    1
6412652    1
6410605    1
6367602    1
6291456    1
Name: ID_Cliente, Length: 438510, dtype: int64

vemos que não, pois possuimos IDs duplicados. Além disso, o tamanho das linhas não corresponde com a quantidade de IDs diferentes. Resta ainda uma pergunta: **os IDs duplicados são de linhas duplicadas?**

In [ ]:
df_clientes_cadastrados.duplicated().sum()

0

Não reotrnou nenhum dado duplicado, o que indica que todas as linhas são diferentes umas das outras.

Mas vamos analisar quais sõa as linhas aonde temos IDs duplicados.analisar

In [ ]:
df_clientes_cadastrados[df_clientes_cadastrados['ID_Cliente'].duplicated()].head()

ID_Cliente Genero  ...  Tamanho_Familia  Anos_empregado
421349     7602432      M  ...              2.0         -1000.7
422660     7052783      H  ...              4.0             7.4
423702     7838075      M  ...              1.0             4.5
424480     7053557      H  ...              3.0             6.2
425306     7023651      M  ...              1.0             3.3

[5 rows x 18 columns]

In [ ]:
# agora vemos a quantidade de linhas nessa condição
df_clientes_cadastrados[df_clientes_cadastrados['ID_Cliente'].duplicated()].shape

(47, 18)

Mas a função `duplicated()` não está retornando todos os IDs duplicados, somentes um para cada duplicada. Podemos mudar isso com o parâmtero `keep = False`

In [ ]:
df_clientes_cadastrados[df_clientes_cadastrados['ID_Cliente'].duplicated(keep=False)].shape

(94, 18)

### Resolvendo o problema: excluindo linhas com IDs repetidos

Essa é apenas uma solução, podem haver outras melhores, estudando melhor o caso. Vamos salvar os IDs repetidos caso precisemos deles mais tarde.

In [ ]:
ids_duplicados = df_clientes_cadastrados[df_clientes_cadastrados['ID_Cliente'].duplicated(keep=False)]['ID_Cliente']
ids_duplicados

421211    7702516
421268    7602432
421349    7602432
421464    7836971
421698    7213374
           ...   
433158    7282535
433159    7742853
433217    7135270
433666    7091721
433789    7618285
Name: ID_Cliente, Length: 94, dtype: int64

In [ ]:
df_clientes_cadastrados_id_unico = df_clientes_cadastrados.drop(ids_duplicados.index)
df_clientes_cadastrados_id_unico.shape

(438463, 18)

In [ ]:
# conferindo a quantidade de IDs únicos
df_clientes_cadastrados_id_unico['ID_Cliente'].value_counts()

6293503    1
5166449    1
6210931    1
5174645    1
5170551    1
          ..
7410037    1
6357367    1
6375806    1
5683584    1
6291456    1
Name: ID_Cliente, Length: 438463, dtype: int64

## Segunda pergunta: existem dados nulos?

In [ ]:
df_clientes_cadastrados_id_unico.isnull().sum().sum()

0

In [ ]:
# por fim, fazendo uma última exploração das features:
df_clientes_cadastrados_id_unico.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 438463 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438463 non-null  int64  
 1   Genero                 438463 non-null  object 
 2   Tem_Carro              438463 non-null  int64  
 3   Tem_Casa_Propria       438463 non-null  int64  
 4   Qtd_Filhos             438463 non-null  int64  
 5   Rendimento_Anual       438463 non-null  float64
 6   Categoria_de_renda     438463 non-null  object 
 7   Grau_Escolaridade      438463 non-null  object 
 8   Estado_Civil           438463 non-null  object 
 9   Moradia                438463 non-null  object 
 10  Idade                  438463 non-null  float64
 11  Tem_Celular            438463 non-null  int64  
 12  Tem_telefone_trabalho  438463 non-null  int64  
 13  Tem_telefone_fixo      438463 non-null  int64  
 14  Tem_email              438463 non-nu

# Desafios

Desafio 1: Explorar IDs repetidos (Atualização de cadastro? Erro?).

Desafio 2: Explorar as colunas (features) identificar continuas, categóricas e binárias.

## Desafio 1: Explorar IDs repetidos (Atualização de cadastro? Erro?).

Vamos explorar o que pode ter causado essa repetição de IDs

In [ ]:
df_clientes_cadastrados[df_clientes_cadastrados['ID_Cliente'].duplicated(keep=False)].query('ID_Cliente == 7702516	')

ID_Cliente Genero  Tem_Carro  ...  Ocupacao  Tamanho_Familia  Anos_empregado
421211     7702516      M          0  ...     Venda              4.0             3.4
432379     7702516      M          0  ...  Medicina              3.0            11.7

[2 rows x 18 columns]

pelo menos o IDs de cliente '7702516' não parece se referir a um mesmo cliente. Podemos ver que o ID remete a 2 clientes com a mesma idade de 32 anos, mas com tempos de empregado bem diferentes. Não parece ser o caso de atualização de cadastro. Talvez possa ser um erro de cadastro.

Vamos analisar mais clientes

In [ ]:
df_clientes_cadastrados[df_clientes_cadastrados['ID_Cliente'].duplicated(keep=False)].sort_values(by='ID_Cliente')

ID_Cliente Genero  ...  Tamanho_Familia  Anos_empregado
426818     7022197      H  ...              5.0             2.0
425023     7022197      M  ...              1.0             4.9
431545     7022327      M  ...              1.0            14.5
431911     7022327      H  ...              2.0             4.6
425486     7023108      H  ...              3.0             4.6
...            ...    ...  ...              ...             ...
426563     7836711      M  ...              4.0            12.0
421464     7836971      H  ...              3.0            15.1
428620     7836971      M  ...              2.0             7.7
422068     7838075      H  ...              2.0             3.5
423702     7838075      M  ...              1.0             4.5

[94 rows x 18 columns]

Realmente os IDs iguais não representam as mesmas pessoas. Há casos de homens e mulheres possuirem o mesmo ID. Além disso há vários outros tipos de dados diferentes.

Uma pergunta a se fazer agora é: existem dados repetidos, de uma mesma pessoa, mas com IDs diferentes?

In [ ]:
df_clientes_cadastrados[df_clientes_cadastrados.drop('ID_Cliente', axis=1).duplicated()].shape

(353636, 18)

In [ ]:
438463 - 353636

84827

A resposta é impressionante! Existem **353636 linhas** com dados repetidos! Ou seja, somente 438463 - 353636 = 84827 clientes possuem dados distintos entre si (sem considerar o ID).



In [ ]:
df_clientes_cadastrados[df_clientes_cadastrados.drop('ID_Cliente', axis=1).duplicated()].head()

ID_Cliente Genero  Tem_Carro  ...  Ocupacao  Tamanho_Familia  Anos_empregado
1     5008805      H          1  ...    Outros              2.0            12.4
4     5008809      M          0  ...     Venda              1.0             8.4
5     5008810      M          0  ...     Venda              1.0             8.4
6     5008811      M          0  ...     Venda              1.0             8.4
8     5008813      M          0  ...    Outros              1.0         -1000.7

[5 rows x 18 columns]

Com essas verificações podemos afirmar que de fato os IDs não são únicos para cada cliente e não podemos confiar neles.

Resta uma pergunta: existem linhas no dataframe **totalmente iguais**?

In [ ]:
df_clientes_cadastrados.duplicated().sum()

0

A resposta é não! Podemos ter a garantia de que realmente todas as linhas são diferentes.

## Desafio 2: Explorar as colunas (features) identificar continuas, categóricas e binárias.

Vamos novamente fazer um `.info()` nos dados

In [ ]:
df_clientes_cadastrados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 438557 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   ID_Cliente             438557 non-null  int64  
 1   Genero                 438557 non-null  object 
 2   Tem_Carro              438557 non-null  int64  
 3   Tem_Casa_Propria       438557 non-null  int64  
 4   Qtd_Filhos             438557 non-null  int64  
 5   Rendimento_Anual       438557 non-null  float64
 6   Categoria_de_renda     438557 non-null  object 
 7   Grau_Escolaridade      438557 non-null  object 
 8   Estado_Civil           438557 non-null  object 
 9   Moradia                438557 non-null  object 
 10  Idade                  438557 non-null  float64
 11  Tem_Celular            438557 non-null  int64  
 12  Tem_telefone_trabalho  438557 non-null  int64  
 13  Tem_telefone_fixo      438557 non-null  int64  
 14  Tem_email              438557 non-nu

In [ ]:
df_clientes_cadastrados_id_unico.head(2)

ID_Cliente Genero  Tem_Carro  ...  Ocupacao  Tamanho_Familia  Anos_empregado
0     5008804      H          1  ...    Outros              2.0            12.4
1     5008805      H          1  ...    Outros              2.0            12.4

[2 rows x 18 columns]

podemos fazer algumas observações:

- por padrão, o pandas entendeu a feature `ID_Cliente` como do tipo inteira, por conta da representação numérica. Mas sabemos que essa informação não possui um quantidade numérica do tipo que possa ser ordenada. Então, é possível que seja melhor alterar essa feature para o tipo caractere.

- as features `Genero`, `Tem_Carro`, `Tem_Casa_Propria`, `Qtd_Filhos`, `Categoria_de_renda`, `Grau_Escolaridade`, `Estado_Civil`, `Moradia`, `Tem_Celular`, `Tem_telefone_trabalho`, `Tem_telefone_fixo`, `Tem_email`, `Ocupacao` e `Tamanho_Familia` são de tipos diferentes. Algumas do tipo caractere, outras do tipo inteiro, mas todas podem ser classificadas como features categóricas.